src/averageAges.js

```diff
@@ -1,64 +1,50 @@
 'use strict';
 
-/**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
+const FEMALE = 'f';
+const MALE = 'm';
+const HUNDRED = 100;
+
+function getAverageAge(people) {
+  return people.reduce((accum, { died, born }) =>
+    (accum + (died - born)), 0) / people.length;
+};
+
+function getAgeDiff(people1, people2) {
+  return ((people1.reduce((accum, { born, mother }) =>
+    accum + born - people2.find(({ name }) => mother
+      === name).born, 0)) / people1.length).toFixed(2);
+};
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const allMen = people.filter(({ sex }) => sex === MALE);
+
+  const men = century
+    ? allMen.filter(({ died }) => Math.ceil(died / HUNDRED) === century)
+    : allMen;
+
+  return getAverageAge(men);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const allWomen = people.filter(({ sex }) => sex === FEMALE);
+
+  const women = withChildren
+    ? allWomen.filter(({ name }) => people.some(({ mother }) =>
+      name === mother))
+    : allWomen;
+
+  return getAverageAge(women);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const children = people.filter(({ mother }) => people.some(({ name }) =>
+    mother === name));
+
+  const sons = children.filter(({ sex }) => sex === MALE);
```

Comment:
When the value of `onlyWithSon` is false, you make an extra filter and worsen the performance, it is better to first check whether this value is true and then do filtering.